### ISSUE TITLE
ScreenCapturer video doesn't handle orientation change

### ISSUE THREAD

**Issue by ryanjohn1** [twilio/video-quickstart-android#320]
Description

I previously included this as a comment on a related issue which is now closed so I'm raising it as it's own issue. From: https://github.com/twilio/video-quickstart-android/issues/287
If you start streaming the screen in one orientation then change, the stream dimensions remain the size of the original and you end up with a tiny display on the receiving device.
I've recorded a gif to demonstrate from the viewers perspective on an iOS device: it's a little difficult to follow as you can't see the device being rotated, so I'll explain the 4 combinations of views being shown:

We start with host in landscape, viewer is portrait. (all good)
Viewer rotates to landscape to match (all good)
Host now rotates to portrait (also correctly showing)
Viewer rotates to landscape (not right)

https://user-images.githubusercontent.com/6177486/39475901-028206a0-4d52-11e8-9964-73d207127d19.gif
So in the final stage you can see that the video stream still has the initial aspect ratio (slightly less dark than the rest of screen) and the portrait video is being shown fitted into landscape stream. It's the same if the viewer is on Android as well. An iOS host works correctly following the same steps and displaying on either our Android or iOS client so I believe the issue is on the sending side, on Android.

Should something be informing Twilio/WebRTC that the device has been rotated and to change the orientation of the video that is being sent? I was wondering if the issue is because I'm running the ScreenCapturer from a foreground service and perhaps rotation events aren't being recognised.

Versions
All relevant version information for issue.

Video Android SDK
2.1.1

Android API
21+

Android Device
Any

**Comment by aaalaniz** [comment]
Hey @ryanjohn1 

I captured this in our backlog. We will take a look. It would be really helpful if you could reproduce the issue in a fork of our quickstart so we can work with a common code base. 

Thanks!

**Comment by zackm0571** [comment]
@ryanjohn1 I'm having trouble reproducing the behavior you're seeing on our end. Could you provide a sample of your layout and any relevant code relating to your capturers and renderers?

Thanks!

**Comment by ryanjohn1** [comment]
@zackm0571 I've managed to reproduce this by modifying the Quickstart example with some code from the ScreenCapturer example. I've got it on a branch of this repo. I don't have permissions to push to the remote so I've just zipped up the modified VideoActivity.java file. 

[VideoActivity.java.zip](https://github.com/twilio/video-quickstart-android/files/2388718/VideoActivity.java.zip)

To reproduce with this code: After accepting the screen share permission, two devices can connect to the same room in the same way as the standard quickstart and their screens will be shown to each other. One device can then turn off sending their screen by tapping the Video floating action button so one device is just viewing the other screen.
The bug is illustrated if both devices then rotate to landscape and you can see that the transmitted screen contents are only a small slice within the video.

Let me know if it's not clear!
Cheers

**Comment by zackm0571** [comment]
@ryanjohn1 thank you so much for the attention to detail here. Will take a look this morning and let you know if I can successfully reproduce.

**Comment by zackm0571** [comment]
@ryanjohn1 I was able to reproduce the issue locally. Thanks again for putting this modified quickstart together. I'll keep you updated as I make progress.

**Comment by zackm0571** [comment]
@ryanjohn1 as silly as this is, by setting the `layout_width` property to `match_parent` on the VideoView with id `primary_video_view` I saw dramatic improvement to the way the remote screencap was being rendered. Could you give this a shot and tell me if you see any improvement? If not we can continue to triage further :)

**Comment by ryanjohn1** [comment]
@zackm0571 we were previously doing that as a way of sort of getting around the problem but all it is really doing is zooming in on the part of the video with content. When it's set to match_parent the VideoScaleType is automatically set to fill (See comment on VideoView setVideoScaleType method), This means the the video is stretched to fit, say you have an initial video that's 1024 tall and 768 pixels wide initially in portrait. You rotate the host device, it keeps the video in the same dimensions so the content will be 768 wide but with a small height as it's landscape. This is put into the original video dimensions with large amounts of black padding (within the video itself, I'm not talking about how it's displayed). If you look carefully you can see the difference between the black pixels within the video to the black pixels of the padding. 

Using match parent as you say does sort of fix it for that one case where both devices have rotated to landscape although it doesn't fit quite correctly and also it will be terrible and massively zoomed in if you try the host rotated but leave the viewing device to be in it's original orientation. Plus the video is low quality because most of the pixels are black and the actual content is just stretched.

I'm pretty sure something should be being done on the host device's side to change the dimensions of the video when the device is rotated. Our iOS host client works fine with our Android viewing code as it is. Our iOS code is similar to this PR https://github.com/twilio/video-quickstart-swift/pull/287

I've tried digging into the ScreenCapturer code a little and noticed that the onTextureFrameCaptured method gets the width, height and rotation angle but this never changes when the device is rotated.
The ScreenCapturerAndroid WebRTC class has a changeCaptureFormat method which I believe would cause it to create a new VirtualDisplay and update the width and height of the frames being captured.
Does anything in the Twilio code call that? I can't see anything.
Perhaps it just needs a new public method on the ScreenCapturer to call through to changeCaptureFormat which we could call once the device has rotated. I can't test if that works as the ScreenCapturerAndroid is private within ScreenCapturer

**Comment by zackm0571** [comment]
@ryanjohn1 thanks for the details. The ScreenCapturerAndroid class does look to be the right place to start. I'll keep you updated on my findings.

**Comment by N-Bark** [comment]
Hi @zackm0571 
I tried to call changeCaptureFormat (using reflection) in the onConfigurationsChanged callback in my streaming service class and it works as expected.
I did a CustomScreenCapturer (as in issue #287) and added below public method in that class.
```  
public void changeCaptureFormat(int width, int height) {
        try {
            Field webRtcField = ScreenCapturer.class.getDeclaredField("webRtcScreenCapturer");
            webRtcField.setAccessible(true);
            ScreenCapturerAndroid screenCapturerAndroid = (ScreenCapturerAndroid)webRtcField.get(this);
            screenCapturerAndroid.changeCaptureFormat(width, height, 30);
        } catch (NoSuchFieldException | IllegalAccessException e) {
            if (DEBUG) {
                throw new RuntimeException("\nReflection failed in CustomScreenCapturer due to:"
                        + "\ne.getMessage():" + e.getMessage());
            } else {
                Log.d(TAG, "Exception:" + e.getMessage());
            }
        }
    }
```
@ryanjohn1 
Is there any progress in this issue? If not, could you please make webRtcScreenCapturer public in the next release (to avoid reflection).

Thanks Niklas

**Comment by zackm0571** [comment]
@ryanjohn1 what type of room are you experiencing the issue in? Group or P2P? If you've only tested one can you test the other? Doing some testing on our end to validate a theory.

**Comment by zackm0571** [comment]
@ryanjohn1 I apologize for the delay. I was able to get a temporary fix by using the following in your code example:

```
 @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        if(newConfig.orientation != orientation){
            orientation = newConfig.orientation;
            if(localParticipant != null) {
                localParticipant.unpublishTrack(localVideoTrack);
                stopScreenCapture();
                startScreenCapture();
                localParticipant.publishTrack(localVideoTrack);
            }
        }
    }
```

Could you please give this a shot for us and see if this works on your end? This is only a work around and we currently have a more robust fix in the works. Thank you for your patience!

**Comment by zackm0571** [comment]
We have released a fix for this in Video Android `4.1.2`. If anyone continues to have issues please feel free to open a new issue. 🙂
